### PR TITLE
feat: Importable/Exportable keys

### PR DESCRIPTION
**Pull request by cristianIOHK**
# Overview
Define Importable/Exportable interfaces and add them to all our public and private keys (Secp256k1, Ed25519, X25519) and use them in Pluto.
It also includes the implementation of StorableKey as it was missing from past work.

Fixes #<issue number>

### My PR contains...
* [ ] No code changes (changes to documentation, CI, metadata, etc.)
* [ ] Bug fixes (non-breaking change which fixes an issue)
* [x] Improvements (misc. changes to existing features)
* [x] Features (non-breaking change which adds functionality)

### My changes...
* [x] are breaking changes
* [ ] are not breaking changes
* [ ] If yes to above: I have updated the documentation accordingly

### Documentation
* [ ] My changes do not require a change to the project documentation
* [x] My changes require a change to the project documentation
* [ ] If yes to above: I have updated the documentation accordingly

### Tests
* [ ] My changes can not or do not need to be tested
* [ ] My changes can and should be tested by unit and/or integration tests
* [ ] If yes to above: I have added tests to cover my changes
* [ ] If yes to above: I have taken care to cover edge cases in my tests

### CODE REVIEW

**Comment by atala-dev** [comment]
### Code Coverage
|File|Coverage [35.24%]|
|:-|:-:|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/Ed25519PrivateKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FEd25519PrivateKey.kt)|35.71%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/Ed25519PublicKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FEd25519PublicKey.kt)|38.46%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/Secp256k1PrivateKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FSecp256k1PrivateKey.kt)|34.48%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/Secp256k1PublicKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FSecp256k1PublicKey.kt)|44.12%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/X25519PrivateKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FX25519PrivateKey.kt)|30.77%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/apollo/utils/X25519PublicKey.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fapollo%2Futils%2FX25519PublicKey.kt)|33.33%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/pluto/PlutoImpl.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fpluto%2FPlutoImpl.kt)|0.00%|
|[atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/prismagent/PrismAgent.kt](https://github.com/input-output-hk/atala-prism-wallet-sdk-kmm/blob/7cbaafde692af836fac284a4698bab9b462363bd/atala-prism-sdk%2Fsrc%2FcommonMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fwalletsdk%2Fprismagent%2FPrismAgent.kt)|53.40%|

|Total Project Coverage|25.82%|
|:-|:-:|

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+data class PEMKey(val keyType: String, val keyData: ByteArray) {
+    constructor(keyType: String, keyData: String) : this(keyType, keyData.base64UrlDecodedBytes)
+
+    fun pemEncoded(): String {
+        val base64Data = keyData.base64UrlEncoded
+        val beginMarker = "-----BEGIN $keyType-----"
+        val endMarker = "-----END $keyType-----"
+
+        return "$beginMarker\n$base64Data$endMarker"
```
Comment (hamada147) [inline atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:51]:
```suggestion
        return "$beginMarker\n$base64Data\n$endMarker"
```

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    val dp: String? = null,
+    val dq: String? = null,
+    val qi: String? = null,
+    val crv: String? = null,
+    val x: String? = null,
+    val y: String? = null,
+    val k: String? = null
+)
+
+data class PEMKey(val keyType: String, val keyData: ByteArray) {
```
Comment (hamada147) [inline atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:43]: Can we change the `keyType` to be an enumeration as we have many types for the keys type, and they must be in a specific format 
ex: `-----BEGIN RSA PRIVATE KEY-----`, `-----BEGIN RSA PUBLIC KEY-----`, `-----BEGIN PRIVATE KEY-----`, `-----BEGIN PUBLIC KEY-----`, etc.

The current implementation is for typical keys (Public and Private keys) without any additional information.
Also, so that users of the `PEMKey` can always get a valid encoded `PEM`.

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            val lines = pemString.split("\n")
+            if (lines.size < 3) {
+                return null
+            }
+
+            val beginMarker = lines[0]
+            val endMarker = lines[lines.size - 1]
+
+            if (!beginMarker.startsWith("-----BEGIN ") || !beginMarker.endsWith("-----") ||
+                !endMarker.startsWith("-----END ") || !endMarker.endsWith("-----")
```
Comment (hamada147) [inline atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:65]: For readability 
```suggestion
            if (beginMarker.startsWith("-----BEGIN ").not() || beginMarker.endsWith("-----").not() ||
                .not()endMarker.startsWith("-----END ") || endMarker.endsWith("-----").not()
```

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    val x: String? = null,
+    val y: String? = null,
+    val k: String? = null
+)
+
+data class PEMKey(val keyType: String, val keyData: ByteArray) {
+    constructor(keyType: String, keyData: String) : this(keyType, keyData.base64UrlDecodedBytes)
+
+    fun pemEncoded(): String {
+        val base64Data = keyData.base64UrlEncoded
```
Comment (hamada147) [inline atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:47]: It should be base64 padded and not the URL version as per the RFC https://www.rfc-editor.org/rfc/rfc7468
Also, the base64 string value should be divided into lines with a max of 64 char per line

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+            if (!beginMarker.startsWith("-----BEGIN ") || !beginMarker.endsWith("-----") ||
+                !endMarker.startsWith("-----END ") || !endMarker.endsWith("-----")
+            ) {
+                return null
+            }
+
+            val keyType = beginMarker.substring(11, beginMarker.length - 5)
+            val base64Data = lines.subList(1, lines.size - 1).joinToString("")
+            val keyData = base64Data.base64UrlDecoded
```
Comment (hamada147) [inline atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:72]: Should be changed to base64 padded not the URL version

**Review by hamada147 — CHANGES_REQUESTED** [review]

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt (hunk 1 above)
Comment (cristianIOHK) [reply atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:51]: Done

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt (hunk 3 above)
Comment (cristianIOHK) [reply atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:65]: Done

File: atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt (hunk 4 above)
Comment (cristianIOHK) [reply atala-prism-sdk/src/commonMain/kotlin/io/iohk/atala/prism/walletsdk/domain/models/keyManagement/ExportableImportableKey.kt:47]: Done

**Comment by cristianIOHK** [comment]
@hamada147 I made all your changes

**Review by hamada147 — APPROVED** [review]
